From cd93f4af6e2940bd3a90afd0b2c45a68ad1e0d9a Mon Sep 17 00:00:00 2001
From: Takashi Yano <takashi.yano@nifty.ne.jp>
Date: Wed, 8 Jul 2026 21:54:45 +0900
Subject: [PATCH 63/N] Cygwin: pty: Fix nat_pipe_owner_pid when gdb runs
 non-cygwin app

Previously, nat_pipe_owner_pid was incorrectly set to 0 when the
inferior of gdb was a non-cygwin app. Due to this bug, repeatedly
running a non-cygwin app under gdb could lead to an unexpected crash.

This occurred because the previous code in setup_for_non_cygwin_app()
set nat_pipe_owner_pid to exec_dwProcessId, which is correct when the
caller is the stub process of the non-cygwin app. exec_dwProcessId is
the PID of the stub process, and dwProcessId is the PID of the spawned
process in the stub process. However, when the caller is gdb, since
the owner should be gdb itself, nat_pipe_owner_pid must be set to
myself->dwProcessId where the normal process's PID is stored.

With this fix, attach_console_temporarily() can be called with target
pid equal to the process's own pid in gdb, in which case the attach
operation is skipped.

Note that this issue commonly occurs in programs that launch non-
cygwin app directly via CreateProcess(), as gdb does.

Fixes: 1e6c51d74136 ("Cygwin: pty: Reorganize the code path of setting up and closing pcon.")
Signed-off-by: Takashi Yano <takashi.yano@nifty.ne.jp>
Reviewed-by: Johannes Schindelin <Johannes.Schindelin@gmx.de>
(cherry picked from commit 50f4ff48727f138bded3f1c4e41357b46df1ef95)
---
 winsup/cygwin/fhandler/pty.cc | 17 +++++++++++++++--
 1 file changed, 15 insertions(+), 2 deletions(-)

diff --git a/winsup/cygwin/fhandler/pty.cc b/winsup/cygwin/fhandler/pty.cc
index f194508..a9f8eb7 100644
--- a/winsup/cygwin/fhandler/pty.cc
+++ b/winsup/cygwin/fhandler/pty.cc
@@ -3793,7 +3793,8 @@ skip_create:
   replace_nat_handles (hpConIn, hpConOut);
 
   if (!process_alive (get_ttyp ()->nat_pipe_owner_pid))
-    get_ttyp ()->nat_pipe_owner_pid = myself->exec_dwProcessId;
+    get_ttyp ()->nat_pipe_owner_pid =
+      myself->exec_dwProcessId ? : myself->dwProcessId;
 
   if (hpcon && nat_pipe_owner_self (get_ttyp ()->nat_pipe_owner_pid))
     {
@@ -4503,7 +4504,12 @@ fhandler_pty_slave::setup_for_non_cygwin_app (bool nopcon,
       fhandler_pty_slave *ptys = (fhandler_pty_slave *) fh;
       ptys->get_ttyp ()->switch_to_nat_pipe = true;
       if (!process_alive (ptys->get_ttyp ()->nat_pipe_owner_pid))
-	ptys->get_ttyp ()->nat_pipe_owner_pid = myself->exec_dwProcessId;
+	/* In normal case where the current process is the stub process for
+	   non-cygwin app, set owner to exec_dwProcessId (which is the PID
+	   of the stub process itself). In gdb case, since gdb itself
+	   should be the owner, the owner pid must be set to dwProcessId. */
+	ptys->get_ttyp ()->nat_pipe_owner_pid =
+	  myself->exec_dwProcessId ? : myself->dwProcessId;
     }
   bool pcon_enabled = false;
   if (!nopcon)
@@ -4631,6 +4637,8 @@ fhandler_pty_common::attach_console_temporarily (DWORD target_pid)
 {
   DWORD resume_pid = 0;
   acquire_attach_mutex (mutex_timeout);
+  if (target_pid == GetCurrentProcessId ())
+    return target_pid;
   pinfo pinfo_resume (myself->ppid);
   if (pinfo_resume)
     resume_pid = pinfo_resume->dwProcessId;
@@ -4649,6 +4657,11 @@ fhandler_pty_common::attach_console_temporarily (DWORD target_pid)
 void
 fhandler_pty_common::resume_from_temporarily_attach (DWORD resume_pid)
 {
+  if (resume_pid == GetCurrentProcessId ())
+    {
+      release_attach_mutex ();
+      return;
+    }
   bool console_exists = (resume_pid != (DWORD) -1);
   if (!console_exists || resume_pid)
     {
